lib/cog/commands/pluck.ex

```diff
@@ -0,0 +1,112 @@
+defmodule Cog.Commands.Pluck do
+  use Spanner.GenCommand.Base, bundle: Cog.embedded_bundle, enforcing: false, calling_convention: :all
+  alias Cog.Command.Helpers, as: CommandHelpers
+
+  @moduledoc """
+  Returns the specified fields from a set of inputs.
+
+  ## Example
+
+      @bot #{Cog.embedded_bundle}:rules --list --for-command="#{Cog.embedded_bundle}:permissions" | #{Cog.embedded_bundle}:pluck --fields=id
+      > { "id": "91edb472-04cf-4bca-ba05-e51b63f26758" }
+        { "id": "738ce922-14cd-4abc-fe94-81658707229c" }
+      @bot #{Cog.embedded_bundle}:rules --list --for-command="#{Cog.embedded_bundle}:permissions" | #{Cog.embedded_bundle}:pluck --fields=rule,command
+      > { "rule": "operable:manage_users",
+          "command": "operable:permissions" }
+        { "rule": "operable:manage_users",
+          "command": "operable:permissions" }
+      @bot #{Cog.embedded_bundle}:seed '[{"foo":{"bar.qux":{"baz":"stuff"}}}, {"foo": {"bar":{"baz":"me"}}}]' | #{Cog.embedded_bundle}:pluck --fields=foo"
+      > [ {"foo": {"bar.qux": {"baz": "stuff"} } }, {"foo": {"bar": {"baz": "me"} } } ]
+      @bot #{Cog.embedded_bundle}:seed '[{"foo":{"bar":{"baz":"stuff"}}}, {"foo": {"bar":{"baz":"me"}}}]' | #{Cog.embedded_bundle}:pluck --fields="foo.bar"
+      > { "foo.bar": {"baz": "stuff"} }
+        { "foo.bar": {"baz": "me"} }
+      @bot #{Cog.embedded_bundle}:seed '[{"foo":{"bar":{"baz":"stuff"}}, "qux": "one"}, {"foo": {"bar":{"baz":"me"}}, "qux": "two"}]' | #{Cog.embedded_bundle}:pluck --fields="qux,\"foo.bar\""
+      > { "qux": "one", "foo.bar": {"baz": "stuff"} }
+        { "qux": "two", "foo.bar": {"baz": "me"} }
+  """
+
+  option "fields", type: "string", required: false
+
+  defstruct req: nil, fields: nil, input: nil, output: nil, errors: []
+
+  def handle_message(req, state) do
+    case req |> validate |> execute |> format do
```

Comment:
Someone recently pointed `with` out to me, and it seems like a pretty useful function for cases like this. You could write this similarly to:
```
...
results = with {:ok, options} <- validate_options(req.options),
                      {:ok, data} <- execute(options, req.args),
                        do: format(data)
case results do
  {:ok, response} ->
    {:reply, req.reply_to, response, state}
  {:error, err} ->
    {:error, req.reply_to, err, state}
end
...
```

Then each of the functions in your flow can just return `{:ok, result}` or `{:error, error}`. `with` will bail on the first error. So you abandon execution at the first sign of trouble. I think it would alleviate the need to pass so much state around. You also don't have to worry about errors from previous functions in functions at the end of the flow. I think it may simplify stuff greatly.

I used a similar approach when I wrote the `which` command. https://github.com/operable/cog/blob/master/lib/cog/commands/which.ex#L31-L42